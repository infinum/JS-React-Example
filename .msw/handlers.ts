import { isJsonApiClass, isModelPersisted, modelToJsonApi } from '@datx/jsonapi';
import { buildMany } from '@datx/test-data-factory';
import { rest } from 'msw';
import { sessionFactory } from '../__factories__/session';
import { userFactory } from '../__factories__/user';
import { flightFactory } from '../__factories__/flight';
import { Session } from '@/models/Session';
import { User } from '@/models/User';

export const handlers = [
	// datx.get({
	// 	op: 'getMany',
	// 	type: Flight.type,
	// } as const, (req, res, ctx) => {

	// }),
	rest.get(`${process.env.NEXT_PUBLIC_API_ENDPOINT}/sessions/current`, (req, res, ctx) => {
		userFactory.reset();
		sessionFactory.reset();
		const include = req.url.searchParams.get('include');
		const session = sessionFactory();

		const included = include
			?.split(',')
			.map((include) => {
				const includeData = session[include];

				if (Array.isArray(includeData)) {
					return includeData.map((item) => modelToJsonApi(item));
				}

				const rawModel = modelToJsonApi(includeData);

				rawModel.id = includeData.id;

				return rawModel;
			})
			.flat();

		const rawSessionModel = modelToJsonApi(session);
		rawSessionModel.id = session.id;

		return res(
			ctx.json({
				data: rawSessionModel,
				included,
			})
		);
	}),
	rest.get(`${process.env.NEXT_PUBLIC_API_ENDPOINT}/flights`, (_req, res, ctx) => {
		flightFactory.reset();
		const data = buildMany(flightFactory, 2);

		return res(
			ctx.json({
				data: data.map((flight) => modelToJsonApi(flight)),
			})
		);
	}),
];
