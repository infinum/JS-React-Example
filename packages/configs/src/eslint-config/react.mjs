import { fixupPluginRules } from '@eslint/compat';
import pluginReact from 'eslint-plugin-react';
import pluginReactHooks from 'eslint-plugin-react-hooks';

export default [
	{
		files: ['**/*.tsx', '**/*.jsx'],
		settings: {
			react: {
				version: 'detect',
			},
		},
		plugins: {
			react: pluginReact,
			'react-hooks': fixupPluginRules(pluginReactHooks),
		},
		rules: {
			...pluginReact.configs['jsx-runtime'].rules,
			...pluginReactHooks.configs.recommended.rules,
		},
	},
];
