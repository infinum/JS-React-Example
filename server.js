const path = require('path');

const nextPath = path.join(__dirname, 'node_modules', '.bin', 'next');

process.argv.length = 1;
process.argv.push(nextPath, 'start');

require(nextPath);
