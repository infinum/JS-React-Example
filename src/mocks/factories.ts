import { createClient } from '@/datx/create-client';
import { Session } from '@/models/Session';
import { User } from '@/models/User';
import { createFactory, oneOf, perBuild, sequence } from '@datx/test-data-factory';

const client = createClient();
const factory = createFactory(client);

export const userFactory = factory(User, {
	fields: {
		id: sequence(),
		firstName: oneOf('Johnny', 'Michael', 'Philip', 'Darius', 'Nieves', 'Christian'),
		lastName: oneOf('Smith', 'Brown', 'Miller'),
	},
});

export const sessionFactory = factory(Session, {
	fields: {
		id: sequence(),
		email: sequence((x) => `user${x}@example.com`),
		user: perBuild(() => userFactory()),
	},
});
