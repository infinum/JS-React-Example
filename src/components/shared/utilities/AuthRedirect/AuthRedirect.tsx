import { FC, useEffect } from 'react';
import { useRouter } from 'next/router';
import { UrlObject } from 'url';

import { useSession } from '@/hooks/use-session';
import { Session } from '@/models/Session';

interface IAuthRedirectProps {
	/**
	 * `href` param passed to Next.js `Router.push` method
	 */
	to: UrlObject | string;
	/**
	 * If this is set to `true` user will be redirected if he is logged in.
	 * Useful when you don't want to show login page to already logged in users.
	 */
	ifFound?: boolean;
	/**
	 * Callback that will trigger a redirect if true is returned.
	 * Useful when you need to redirect base on some attribute, e.g. if user is not admin
	 */
	condition?(session: Session): boolean;
}

export const AuthRedirect: FC<IAuthRedirectProps> = ({ to, ifFound, condition }) => {
	const { data, isValidating, error } = useSession();
	const router = useRouter();

	useEffect(() => {
		// if state is validating, wait until request is done
		if (isValidating) {
			return;
		}

		// https://swr.vercel.app/advanced/performance#dependency-collection
		const hydration = data === undefined && error === undefined && isValidating === false;

		// if user data not yet there (fetch in progress, logged in or not) then don't do anything yet
		if (hydration) {
			return;
		}

		// `condition` has a priority over a `ifFound` property
		const shouldRedirect = condition ? condition(data.data) : (ifFound && data) || (!ifFound && !data);

		if (shouldRedirect) {
			router.push(to);
		}
	}, [data, ifFound, to, error, isValidating, router, condition]);

	// this component renders nothing since it is only used to redirect if needed
	return null;
};
