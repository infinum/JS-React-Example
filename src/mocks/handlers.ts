import { buildJsonApiDocument } from '@/datx/buildJsonApiDocument';
import { TEST_INVALID_EMAIL, TEST_INVALID_PASSWORD } from '__mocks__/constants';
import { sessionFactory } from '__mocks__/factories';
import { rest } from 'msw';

export const generateTestApiUrl = (path: string) => {
	return process.env.API_TEST_ENDPOINT + path;
};

export const MOCKED_URLS = {
	SessionCurrent: generateTestApiUrl('sessions/current'),
	SessionMe: generateTestApiUrl('sessions/me'),
	Session: generateTestApiUrl('sessions'),
	Flights: generateTestApiUrl('flights'),
} as const;

export const handlers = [
	rest.get(MOCKED_URLS.SessionCurrent, async (_req, res, ctx) => {
		return res(
			ctx.status(401),
			ctx.json({
				errors: [
					{
						status: 'unauthorized',
						code: 'UNAUTHORIZED',
						title: 'Unauthorized',
						detail: 'Must be logged in to perform this action',
					},
				],
			})
		);
	}),
	rest.post(MOCKED_URLS.Session, async (req, res, ctx) => {
		const body = await req.json();
		const emailOverride = body.data?.attributes?.email || undefined;

		return res(
			ctx.status(200),
			ctx.json(
				sessionFactory({
					overrides: {
						email: emailOverride,
					},
				})
			)
		);
	}),
	rest.delete(MOCKED_URLS.SessionMe, async (_req, res, ctx) => {
		return res(ctx.status(204));
	}),
];

export const handlerOverrides = {
	invalidEmailLogin: rest.post(MOCKED_URLS.Session, (_req, res, ctx) => {
		return res(
			ctx.status(422),
			ctx.json({
				errors: [
					{
						status: 'unprocessable entity',
						code: 'UNPROCESSABLE_ENTITY',
						title: 'Unprocessable Entity',
						detail: TEST_INVALID_EMAIL,
						source: { parameter: 'email', pointer: 'data/attributes/email' },
					},
				],
			})
		);
	}),
	invalidPasswordLogin: rest.post(MOCKED_URLS.Session, (_req, res, ctx) => {
		return res(
			ctx.status(422),
			ctx.json({
				errors: [
					{
						status: 'unprocessable entity',
						code: 'UNPROCESSABLE_ENTITY',
						title: 'Unprocessable Entity',
						detail: TEST_INVALID_PASSWORD,
						source: { parameter: 'password', pointer: 'data/attributes/password' },
					},
				],
			})
		);
	}),
	activeCurrentSession: rest.get(MOCKED_URLS.SessionCurrent, async (req, res, ctx) => {
		const session = sessionFactory();

		return res(ctx.status(200), ctx.json(buildJsonApiDocument(session)));
	}),
};
