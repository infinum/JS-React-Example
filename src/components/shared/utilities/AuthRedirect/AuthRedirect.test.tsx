import { AuthRedirect, IAuthRedirectProps } from '@/components/shared/utilities/AuthRedirect/AuthRedirect';
import { FC } from 'react';
import { render, waitFor } from 'test-utils';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';
import mockRouter from 'next-router-mock';
import { server } from '@/mocks/server';
import { MOCKED_URLS, handlerOverrides } from '@/mocks/handlers';
import { rest } from 'msw';

const TestAuthRedirect: FC<IAuthRedirectProps> = ({ to = '/', ifFound = true }) => (
	<MemoryRouterProvider>
		<AuthRedirect to={to} ifFound={ifFound} />
	</MemoryRouterProvider>
);

describe('AuthRedirect', () => {
	it('should matches snapshot', () => {
		const { asFragment } = render(<TestAuthRedirect to="/" ifFound />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should redirect logged in user', async () => {
		server.use(handlerOverrides.activeCurrentSession);

		mockRouter.push('/login');

		render(<TestAuthRedirect to="/" ifFound />);

		await waitFor(() => {
			expect(mockRouter.asPath).toEqual('/');
		});
	});

	it('should redirect based on condition', async () => {
		server.use(
			rest.get(MOCKED_URLS.SessionCurrent, (req, res, ctx) => {
				return res(
					ctx.status(200),
					ctx.json({
						data: {
							id: 'current',
							type: 'session',
							attributes: {
								email: 'admin@infinum.com',
							},
						},
					})
				);
			})
		);

		mockRouter.push('/login');

		render(<TestAuthRedirect to="/admin" condition={(session) => session?.email === 'admin@infinum.com'} />);

		await waitFor(() => {
			expect(mockRouter.asPath).toEqual('/admin');
		});
	});
});
