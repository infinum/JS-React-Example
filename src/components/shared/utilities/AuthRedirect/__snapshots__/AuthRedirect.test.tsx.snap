// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthRedirect should matches snapshot 1`] = `
<DocumentFragment>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
